### DISCUSSION TITLE
How to prevent being logged-in when trying to link second OAuth provider?

### DISCUSSION THREAD

**Discussion by ArthurMJ96** [pocketbase/pocketbase#3984] [Q&A] [answered]
I'm encountering an issue with OAuth2 account linking in PocketBase. 
Users on my app can create accounts via OAuth2 providers such as Twitch or Google. 
I'm trying to allow a logged-in user to link their Twitch account with their existing account in case they initially created the account via Google. 
The problem is that if another account already exists and is linked with that Twitch account, the PocketBase JavaScript SDK logs the user into that account instead of returning an error.

Perhaps, there's a dedicated linking method? Since there's one for unlinking (`unlinkExternalAuth`)
```ts
const linkTwitchAccount = async () => {
  const res = await $pb
    .collection('users')
    .authWithOAuth2({ provider: 'twitch' })
}
```

**Comment by ganigeorgiev** [comment]
I don't understand what do you mean.

> The problem is that if another account already exists and is linked with that Twitch account, the PocketBase JavaScript SDK logs the user into that account instead of returning an error.

Why is this a problem and why it should return an error?

> Perhaps, there's a dedicated linking method?

The linking happens automatically. We can't have a dedicated linking endpoint, at least not without requiring admin permissions or performing some custom verification based on the OAuth2 token.

Could you elaborate on your auth flow and the desired behavior?

**Reply by ganigeorgiev** [reply]
To clarify, when authenticating with OAuth2:

1. We check first for existing auth record `ExternalAuth` relation. If it exists, authenticate as the linked auth record from the relation. Otherwise fallback to 2.

2. If you send the `authWithOAuth2` request while being already authenticated (aka. `pb.authStore.token` was valid before sending the request) and the OAuth2 user is not already linked to another PocketBase user, then will link the OAuth2 user to the authenticated auth record (_If you want to prevent this step, then it should be enough to call `pb.authStore.clear()` before sending the `authWthOAuth2` request_). Otherwise fallback to 3.

3. If the OAuth2 user has an email, look for PocketBase user with that email. If it exists link it.

4. If the above steps fail, create a new PocketBase user and link it to the OAuth2 one.

**Reply by ArthurMJ96** [reply]
> Could you elaborate on your auth flow and the desired behavior?

I'm sending a `authWithOAuth2` request while being already authenticated to achieve "Step 2". But if it exists I would like to show the user a error message stating that the given account could not be linked because it's already linked with another account.
What is happening instead, is that "Step 1" passes and I'm suddenly authenticated as another user.

**Reply by ganigeorgiev** [reply]
I still don't understand why do you want to get an error.

If you don't want to reach step 2, call `pb.authStore.clear()` before sending the OAuth2 request.

This is technically the same as calling twice `authWithPassword`:

```js
await pb.collection("users").authWithPassword("test1", "1234567890")
await pb.collection("users").authWithPassword("test2", "1234567890") // should this throw?
```

Note that in earlier version of PocketBase we actually had a requirement the auth endpoints to be guest only but several users complained that there were cases where they wanted to have information who send the request and so I decided to remove those checks.

**Reply by ArthurMJ96** [reply]
> I still don't understand why do you want to get an error.

Doesn't need to be an error, but I need an option to know that the given OAuth2-Account couldn't be linked to the provided account (`pb.authStore.token`). And not change my current `pb.authStore` to whatever other account has been found.

> If you don't want to reach step 2, call pb.authStore.clear() before sending the OAuth2 request.

I want to reach step 2, but there is a possibility that the OAuth2-Account I'm trying to link is already linked to another account.

**Reply by ganigeorgiev** [reply]
Ok, I think we are going in circles. I asked "Why" several times but you refuse to elaborate so I'll make one last attempt to explain it and will stop engaging further.

PocketBase APIs are stateless. There are no sessions and each request is executing independently from the state of another. The only thing that keep track of who is authenticated is the `Authorization` header send with each request. This header is added automatically with the requests if `pb.authStore.token` is not empty.

Techincally calling:

```js
await pb.collection("users").authWithPassword("test1", "1234567890")

await pb.collection("users").authWithOAuth2({ provider: "twitch" })
``` 

Is similar to calling:

```js
await pb.collection("users").authWithPassword("test1", "1234567890")

await pb.collection("users").authWithPassword("test2", "1234567890")
```

In other words, the last authentication is a successful authentication. It doesn't matter what was the previous `pb.authStore` state as it will get replaced with the latest one.

**Reply by ArthurMJ96** [reply]
> Why is this a problem and why it should return an error?

It's a problem because there is currently no other way to link one or more OAuth2-Provider-Accounts to an existing PocketBase-Account other then going through the authentication endpoint/api.
In the scenario where authentication is the goal, it shouldn't return an error. 
In the scenario where the goal is to link an OAuth2-Provider-Account to an existing PocketBase-Account, it should return an error. Because the link wasn't successful and with the current implementation we have instead been authenticated into another unrelated PocketBase-Account.



Thank you for explaining the stateless nature of PocketBase APIs. I understand that each request is independent, but in our application flow, we aim to provide users with clear feedback. When attempting to link a Twitch account that is already linked to another PocketBase account, we want to avoid automatic login and instead inform the user with an error message.

This is important for our user experience and user expectations. Is there a recommended approach within the PocketBase SDK to handle this scenario without automatically replacing the current pb.authStore state?

**Reply by ganigeorgiev** [reply]
Fine, this really will be the one last from me.

> When attempting to link a Twitch account that is already linked to another PocketBase account

How this even would happen in the first place?
What you are saying is that User1 somehow has access to the Twitch credentials of User2.
This is no different than User1 having the password for User2.

In any case if you want to authenticate without modifying your current `pb.authStore` state, you can create a one-off SDK instance and authenticate with it.

```js
import PocketBase, { LocalAuthStore } from "pocketbase"

const pb = new PocketBase("...")

...

// create a new temp client with a custom localStorage key
const tempClient = new PocketBase(pb.baseUrl, new LocalAuthStore("oauth2_auth"));
tempClient.authStore.save(pb.authStore.token); // allow linking (if possible)
await tempClient.collection("users").authWithOAuth2({ provider: "twitch" })

if (tempClient.authStore.model.id != pb.authStore.model?.id) {
    throw new Error("Logged as another user...");
}
```

**Edit: added a new instance of the `LocalAuthStore` with different key because otherwise the localStorage state will be shared.**

**Reply by ArthurMJ96** [reply, marked as answer]
It's a frequent occurrence for users to unintentionally create multiple accounts over time when a website provides instant account creation through various providers like Google, Apple, or Twitch. 
This becomes problematic when users attempt to link an additional account, such as Twitch, to their main account for enhanced functionality, only to encounter an error message stating, 'This Twitch account is already used by another account.'

Users, being human, can sometimes make mistakes, and it's the developer's responsibility to anticipate and address these issues. 

With the current state of the API (v0.20.1), linking an additional account is only possible by performing a `authWithOAuth2()` call on top of an already authenticated user. [Example](https://github.com/pocketbase/pocketbase/discussions/1353#discussioncomment-4484954)  
This leads to an unexpected outcome for the end user where, upon encountering the mentioned problem, the logged-in user is logged out and logged into one of the accidental accounts. No indication that the linking was unsuccessful. 

Thank you for providing the workaround. With it, I can reinstate expected behavior by detecting an unsuccessful link attempt.


```ts

import PocketBase, { LocalAuthStore } from "pocketbase"

const pb = new PocketBase('...')

/**
 * Attempts to link the current user with the specified provider.
 * @param provider Name of the provider to link (Can be retrieved from pb.collection('users').listAuthMethods())
 * @returns true if the provider was linked successfully, false otherwise
 */
async function linkProvider (provider: string) {
  const tempClient = new PocketBase(
    pb.baseUrl,
    new LocalAuthStore('link_auth')
  )

  tempClient.authStore.save(pb.authStore.token)
  await tempClient.collection('users').authWithOAuth2({ provider })
  return tempClient.authStore.model?.id == pb.authStore.model?.id
}

// Example usage
const linkTwitch = async () => {
  const success = await linkProvider('twitch')
  alert(`Twitch linked: ${success}`)
}
```
